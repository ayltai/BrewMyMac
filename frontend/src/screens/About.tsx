import mixpanel from 'mixpanel-browser';
import { Box, styled, Typography, } from '@mui/material';
import React, { useCallback, useEffect, } from 'react';
import { useTranslation, } from 'react-i18next';

import { ExternalLink, PopUpDialog, } from '../components';

const LineBreakingTypography = styled(Typography)({
    whiteSpace: 'pre-line',
});

export const About = ({
    onClose,
} : {
    onClose? : () => void,
}) => {
    const { t, } = useTranslation();

    const handleLinkedInClick = useCallback(() => {
        mixpanel.track('Referral - LinkedIn');

        window.open(t<string>('url_linkedin'), '_blank');
    }, []);

    useEffect(() => {
        mixpanel.track('Page View - About');
    }, []);

    return (
        <PopUpDialog
            open
            title={t('label_about_title')}
            onClose={onClose}>
            <Box
                width='100%'
                display='flex'
                justifyContent='center'>
                <img
                    width={128}
                    height={128}
                    src={require('../LogoLarge.png')}
                    alt={t('app_name')} />
            </Box>
            <LineBreakingTypography
                gutterBottom
                marginTop={2}
                variant="body1">
                {t('label_about_description')}
            </LineBreakingTypography>
            <ExternalLink
                title={t('action_follow')}
                onClick={handleLinkedInClick} />
        </PopUpDialog>
    );
};
