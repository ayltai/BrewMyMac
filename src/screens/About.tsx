import mixpanel from 'mixpanel-browser';
import { Box, Link, styled, Typography, } from '@mui/material';
import React, { useEffect, } from 'react';
import { useTranslation, } from 'react-i18next';

import { PopUpDialog, } from '../components';

const ClickableLink = styled(Link)`
    cursor : pointer;
`;

const LineBreakingTypography = styled(Typography)({
    whiteSpace: 'pre-line',
});

export const About = ({
    onClose,
} : {
    onClose? : () => void,
}) => {
    const { t, } = useTranslation();

    const handleLinkedInClick = () => {
        mixpanel.track('Referral', {
            name : 'LinkedIn',
        });

        window.open(t('url_linkedin'), '_blank');
    };

    useEffect(() => {
        mixpanel.track('Page View', {
            name : 'About',
        });
    }, []);

    return (
        <PopUpDialog
            open
            title={t('label_about_title')}
            onClose={onClose}>
            <Box
                width='100%'
                display='flex'
                justifyContent='center'>
                <img
                    width={128}
                    height={128}
                    src={require('../LogoLarge.png')} />
            </Box>
            <LineBreakingTypography
                gutterBottom
                marginTop={2}
                variant="body1">
                {t('label_about_description')}
            </LineBreakingTypography>
            <ClickableLink onClick={handleLinkedInClick}>
                {t('action_follow')}
            </ClickableLink>
        </PopUpDialog>
    );
};
