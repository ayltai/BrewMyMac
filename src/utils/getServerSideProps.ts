import { serverSideTranslations, } from 'next-i18next/serverSideTranslations';
import nextI18NextConfig from '../../next-i18next.config';

export default async function getServerSideProps({
    locale,
} : {
    locale : string,
}) {
    return {
        props : {
            ...(await serverSideTranslations(locale, [
                'common',
            ], nextI18NextConfig)),
        },
    };
}
