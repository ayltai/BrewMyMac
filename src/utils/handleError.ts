import * as Sentry from '@sentry/nextjs';

export const handleError = (error : any) => {
    if (process.env.NODE_ENV === 'production') {
        Sentry.captureException(error);
    } else {
        console.error(error);
    }
};
