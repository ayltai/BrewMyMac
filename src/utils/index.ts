export { createEmotionCache, } from './createEmotionCache';
export { handleError, } from './handleError';
