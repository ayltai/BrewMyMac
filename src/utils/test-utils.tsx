import { createTheme, ThemeProvider, } from '@mui/material/styles';
import { configureStore, PreloadedState, Store, } from '@reduxjs/toolkit';
import '@testing-library/jest-dom/extend-expect';
import { act, render, } from '@testing-library/react';
import React, { ReactNode, } from 'react';
import { Provider, } from 'react-redux';

import { homebrewApi, } from '../apis';
import { preferencesReducer, shoppingCartReducer, store as reduxStore, } from '../states';
import { palette, } from '../styles';

import type { HomebrewFormula, } from '../types';

const INITIAL_FILTER = 'firefox';

export const mockApis = async () => {
    const HOMEBREW_API_RESPONSE : HomebrewFormula[] = [
        {
            token     : `${INITIAL_FILTER} homebrew token`,
            name      : [
                `${INITIAL_FILTER} homebrew name`,
            ],
            full_name : `${INITIAL_FILTER} homebrew full name`,
            bottle    : {},
        },
    ];

    const HOMEBREW_CAST_API_RESPONSE : HomebrewFormula[] = [
        {
            token     : `${INITIAL_FILTER} homebrew cask token`,
            name      : [
                `${INITIAL_FILTER} homebrew cask name`,
            ],
            full_name : `${INITIAL_FILTER} homebrew cask full name`,
            bottle    : {},
        },
    ];

    fetchMock.doMock(request => {
        if (request.url.startsWith('https://formulae.brew.sh/api/formula.json')) {
            return Promise.resolve({
                status : 200,
                body   : JSON.stringify(HOMEBREW_API_RESPONSE),
            });
        }

        if (request.url.startsWith('https://formulae.brew.sh/api/cask.json')) {
            return Promise.resolve({
                status : 200,
                body   : JSON.stringify(HOMEBREW_CAST_API_RESPONSE),
            });
        }

        return Promise.resolve({
            status : 404,
        });
    });

    await act(async () => {
        await reduxStore.dispatch(homebrewApi.endpoints.formula.initiate());
        await reduxStore.dispatch(homebrewApi.endpoints.cask.initiate());
    });
};

const createStore = (preloadedState? : Record<string, any>) => configureStore({
    reducer : {
        preferences  : preferencesReducer,
        shoppingCart : shoppingCartReducer,
    },
    preloadedState,
});

const defaultStore = createStore();

type RootState = ReturnType<typeof defaultStore.getState>;

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : PreloadedState<RootState>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => {
        const theme = createTheme({
            palette : {
                ...palette,
                mode : 'light',
            },
        });

        return (
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    {children}
                </ThemeProvider>
            </Provider>
        );
    },
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
