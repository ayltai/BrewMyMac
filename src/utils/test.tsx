import { createTheme, ThemeProvider, } from '@mui/material';
import { configureStore, PreloadedState, Store, } from '@reduxjs/toolkit';
import '@testing-library/jest-dom/extend-expect';
import { act, render, } from '@testing-library/react';
import React, { ReactNode, } from 'react';
import { Provider, } from 'react-redux';

import { appStoreApi, homebrewApi, tweakApi, } from '../apis';
import type { AppStoreSearchResults, HomebrewFormula, Tweak, } from '../models';
import { sessionReducer, store as reduxStore, } from '../redux';
import { palette, } from '../styles';

const INITIAL_FILTER = 'firefox';

export const mockApis = async () => {
    const HOMEBREW_API_RESPONSE : HomebrewFormula[] = [
        {
            token     : `${INITIAL_FILTER} homebrew token`,
            name      : [
                `${INITIAL_FILTER} homebrew name`,
            ],
            full_name : `${INITIAL_FILTER} homebrew full name`,
        },
    ];

    const HOMEBREW_CAST_API_RESPONSE : HomebrewFormula[] = [
        {
            token     : `${INITIAL_FILTER} homebrew cask token`,
            name      : [
                `${INITIAL_FILTER} homebrew cask name`,
            ],
            full_name : `${INITIAL_FILTER} homebrew cask full name`,
        },
    ];

    const TWEAKS_API_RESPONSE : Tweak[] = [
        {
            id   : '2',
            name : `${INITIAL_FILTER} tweak name`,
        },
    ];

    const APP_STORE_API_RESPONSE : AppStoreSearchResults = {
        resultCount : 1,
        results     : [
            {
                trackId   : '1',
                trackName : `${INITIAL_FILTER} app store name`,
            },
        ],
    };

    fetchMock.doMock(request => {
        if (request.url.startsWith('https://formulae.brew.sh/api/formula.json')) {
            return Promise.resolve({
                status : 200,
                body   : JSON.stringify(HOMEBREW_API_RESPONSE),
            });
        }

        if (request.url.startsWith('https://formulae.brew.sh/api/cask.json')) {
            return Promise.resolve({
                status : 200,
                body   : JSON.stringify(HOMEBREW_CAST_API_RESPONSE),
            });
        }

        if (request.url.startsWith('https://raw.githubusercontent.com')) {
            return Promise.resolve({
                status : 200,
                body   : JSON.stringify(TWEAKS_API_RESPONSE),
            });
        }

        if (request.url.startsWith('https://search-itunes.vercel.app')) {
            return Promise.resolve({
                status : 200,
                body   : JSON.stringify(APP_STORE_API_RESPONSE),
            });
        }

        return Promise.resolve({
            status : 404,
        });
    });

    await act(async () => {
        await reduxStore.dispatch(homebrewApi.endpoints.formula.initiate());
        await reduxStore.dispatch(homebrewApi.endpoints.cask.initiate());
        await reduxStore.dispatch(tweakApi.endpoints.getTweaks.initiate());
        await reduxStore.dispatch(appStoreApi.endpoints.search.initiate(`${INITIAL_FILTER} app store name`));
    });
};

const createStore = (preloadedState? : Record<string, any>) => configureStore({
    reducer : {
        session : sessionReducer,
    },
    preloadedState,
});

const defaultStore = createStore();

type RootState = ReturnType<typeof defaultStore.getState>;

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : PreloadedState<RootState>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => {
        const theme = createTheme({
            palette : {
                ...palette,
                mode : 'light',
            },
        });

        return (
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    {children}
                </ThemeProvider>
            </Provider>
        );
    },
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
