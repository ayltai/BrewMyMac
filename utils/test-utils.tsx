import { createTheme, ThemeProvider, } from '@mui/material/styles';
import { configureStore, PreloadedState, Store, } from '@reduxjs/toolkit';
import '@testing-library/jest-dom/extend-expect';
import { render, } from '@testing-library/react';
import React, { ReactNode, } from 'react';
import { Provider, } from 'react-redux';

import { preferencesReducer, } from '../states';
import { palette, } from '../styles';

const createStore = (preloadedState? : Record<string, any>) => configureStore({
    reducer : {
        preferences : preferencesReducer,
    },
    preloadedState,
});

const defaultStore = createStore();

type RootState = ReturnType<typeof defaultStore.getState>;

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : PreloadedState<RootState>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => {
        const theme = createTheme({
            palette : {
                ...palette,
                mode : 'light',
            },
        });

        return (
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    {children}
                </ThemeProvider>
            </Provider>
        );
    },
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
