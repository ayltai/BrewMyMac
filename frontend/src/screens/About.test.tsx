import React from 'react';

import { render, } from '../utils/test';

import { About, } from './About';

describe('<About />', () => {
    it('renders correctly', () => expect(render(<About />).asFragment()).toMatchSnapshot());
});
