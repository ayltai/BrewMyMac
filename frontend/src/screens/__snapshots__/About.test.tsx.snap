// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<About /> renders correctly 1`] = `<DocumentFragment />`;
